core/runtime/common/runtime_instances_pool.cpp

```diff
@@ -64,52 +63,80 @@ namespace kagome::runtime {
 
    private:
     std::weak_ptr<RuntimeInstancesPool> pool_;
-    RuntimeInstancesPool::RootHash state_;
+    common::Hash256 hash_;  // either trie hash or code hash
     std::shared_ptr<ModuleInstance> instance_;
   };
 
-  RuntimeInstancesPool::RuntimeInstancesPool() : pools_{MODULES_CACHE_SIZE} {}
-
   RuntimeInstancesPool::RuntimeInstancesPool(
       std::shared_ptr<ModuleFactory> module_factory, size_t capacity)
       : module_factory_{std::move(module_factory)}, pools_{capacity} {}
 
-  outcome::result<std::shared_ptr<ModuleInstance>>
-  RuntimeInstancesPool::instantiate(const RootHash &code_hash,
-                                    common::BufferView code_zstd) {
-    std::unique_lock lock{mt_};
-    auto entry = pools_.get(code_hash);
-    if (not entry) {
+  outcome::result<std::shared_ptr<ModuleInstance>, CompilationError>
+  RuntimeInstancesPool::instantiateFromCode(const CodeHash &code_hash,
+                                            common::BufferView code_zstd) {
+    std::unique_lock lock{pools_mtx_};
+    auto pool_opt = pools_.get(code_hash);
+
+    if (!pool_opt) {
       lock.unlock();
-      common::Buffer code;
-      OUTCOME_TRY(uncompressCodeIfNeeded(code_zstd, code));
-      OUTCOME_TRY(module, module_factory_->make(code));
+      OUTCOME_TRY(module, tryCompileModule(code_hash, code_zstd));
+      BOOST_ASSERT(module != nullptr);
       lock.lock();
-      entry = pools_.get(code_hash);
-      if (not entry) {
-        entry = pools_.put(code_hash, {std::move(module), {}});
-      }
+      pool_opt = std::ref(pools_.put(code_hash, InstancePool{module, {}}));
```

Comment:
Insert once (`pools_.get` again) to keep cached instances.

Only first call to `tryCompileModule` (who inserts future) must try to `pools_.put` (after checking `pools_.get`).

Reply:
I'm not sure I get it, could you elaborate?